addons/snailmail/static/tests/chatter_snailmail_tests.js

```diff
@@ -0,0 +1,499 @@
+
+    var form = createView({
+        View: FormView,
+        model: 'account.invoice',
+        res_id: 1,
+        data: this.data,
+        services: this.services,
+        arch: getArch(),
+        mockRPC: function (route, args) {
+            if (args.model === 'snailmail.letter') {
+                assert.step(args.method);
+                if ((args.method === 'cancel_letter' && args.args[0] === 11) || args.method === 'send_letter') {
+                    return $.when();
+                }
+            }
+            if (args.method === 'get_credits_url' && args.model === 'iap.account') {
+                return $.when('credits_url');
+            }
+            return this._super.apply(this, arguments);
+        }
+    });
+
+    assert.ok(form.$('.o_thread_snailmail_tooltip_container').length,
+        "Snailmail icon should appear on message");
+    assert.ok(form.$('.o_thread_message_snailmail_trial_error').length,
+        "Snailmail status of message should be 'trial_error'");
+    assert.notOk(form.$('.o_thread_tooltip_snailmail').length,
+        "No tooltip should be present");
+
+    form.$('.o_thread_message_snailmail_trial_error').trigger('mouseenter');
+    assert.ok($('.o_thread_tooltip_snailmail:visible').length,
+        "Tooltip should appear when hovering the Snailmail Icon");
+    assert.ok($('.o_thread_tooltip_snailmail_icon.fa-exclamation').length,
+        "Tooltip should show correct icon");
+    assert.ok($('.o_thread_tooltip_snailmail:contains("Error")').length,
+        "Tooltip should show correct text");
+
+    testUtils.dom.click(form.$('.o_thread_message_snailmail_trial_error'));
+    var $modal = $('.modal');
+    assert.ok($modal.length,
+        "A modal should open on click");
+
+    var $resendButton = $modal.find('button:contains("Re-send letter")');
+    assert.ok($resendButton.length,
+        "Modal should have a 'Re-send letter' button");
+    testUtils.dom.click($resendButton);
+    assert.notOk($('.modal').length,
+        "The modal should be closed after click on 'Re-send letter'");
+
+    testUtils.dom.click(form.$('.o_thread_message_snailmail_trial_error'));
+
+    var $cancelButton = $('.modal').find('button:contains("Cancel letter")');
+    assert.ok($cancelButton.length,
+        "Modal should have a 'Cancel letter' button");
+    testUtils.dom.click($cancelButton);
+    assert.notOk($('.modal').length,
+        "The modal should be closed after click on 'Cancel letter'");
+
+    assert.verifySteps(['send_letter', 'cancel_letter'], 
+        "Should have made RPC calls to 'send_letter' and 'cancel_letter'");
+
+    form.destroy();
+});
+
+QUnit.test('Missing Required Fields', function (assert) {
+    assert.expect(7);
+
+    this.data['mail.message'].records[0].snailmail_status = 'missing_required_fields';
+
+    testUtils.mock.patch(ThreadWidget, {
+        do_action: function (action, options) {
+            assert.ok(true, "'do_action' should be called");
```

Comment:
prefer `assert.step` and `assert.verifySteps`, it's easier to see error message is a missing step than a wrong assertion count